### ISSUE TITLE
mp.osd_message feature request

### ISSUE THREAD

**Issue by Corginyan** [mpv-player/mpv#7031]
I'm new in scripting for mpv, and use `mp.osd_message` in my scripts in javascript and every time i execute `mp.osd_message` previous message is cleared. So it will be cool if previous message is stay on screen based on it's timeout, and next will be shown in new line, so scripts will not interrupt each other messages. And even if user made multiple actions catched by one separate script - he can view all messages and not only last one. I can use internal loop for it with array of texts with timeouts and hack with `mp.osd_message(resultTextWithNewLines, 999)` and `mp.osd_message("", 0)`, but it looks not good and will break other scripts output. Maybe i still don't get some best practice of scripting for mpv. Sorry for my bad english. By the way, awesome api, thanks for javascript support.

**Comment by wm4** [comment]
Not necessarily a bad idea, but the way OSD is used, this would quickly cause messages to "pile up". Consider if you for example change brightness (keys 3 and 4). This would generate a lot of messages.

So there needs to be a concept to handle this.

**Comment by occivink** [comment]
Maybe you could pass a string id to `osd_message`, and messages with the same id would replace one another, but different ids would be displayed above or below until they timeout.

**Comment by avih** [comment]
Hmm.. or maybe it could be a command prefix, similar to `no-osd` and `osd-message`, maybe there could also be `osd-scroll` or `osd-log` or some such?

**Comment by wm4** [comment]
Yeah, but then it gets complicated very fast. Though I still feel like something to handle this might be a good idea probably.

**Comment by Argon-** [comment]
Well, setting some string permanently and non-overwritable is already possible with `mp.set_osd_ass`, it's just not documented. Maybe this could be the chance to do so.
(https://github.com/mpv-player/mpv/blob/master/player/lua.c#L996 three arguments: int x coord, int y coord, string text)

Imo, the basic question is whether a script or mpv is responsible for managing drawn ass text/layers and using `set_osd_ass` it's the responsibility of a script, which is fine I think.

If I were to change/wish for something, I'd add an additional parameter to `set_osd_ass` so you can draw multiple layers from one script. However, then people will ask for functionality to layout these layers retroactively (e.g. change order), so not sure if that's a good idea...
The end result would be full blown GUI toolkits.

**Comment by Corginyan** [comment]
@Argon- that's really cool, i can draw rectangles, i can even make game like pacman in my videoplayer. Why that calls still not documented i'm really interested. But all i want to is _not_ use internal script loops to draw simple messages, anyway that _ass_ messages will stay same with another scripts _ass_ messages above each other. I don't want to control they live and die, i don't want to set custom position, i just want to show it and forget, but know that user can read them.

**Comment by vorou** [comment]
looks like it's still not documented

for reference, [here's](https://github.com/aerobounce/trim.lua/blob/473c9b1a9900c617ee7bb78521906090b15868d4/trim.lua#L235) an example of how you can write a permanent message

[more examples of assdraw usage](https://github.com/search?q=require+%22mp.assdraw%22&type=code)

**Comment by Argon-** [comment]
Just for reference: it's not a mistake that this still isn't documented. This api was never supposed to be public and was only developed specificaly for the OSC and it's needs. If different scripts start to independently use (parmanent) layers, they will inevitably overlap and there is no layouting mechanism.
